File: deepseek_ocr/hllset/HLLSets.py
"""
HLLSets.py: A Python wrapper for the HllSets.jl Julia module.

This module provides a Pythonic interface to the high-performance HyperLogLog
implementation in Julia. It uses the `juliacall` package to bridge the two
languages.

Prerequisites:
1. Julia must be installed and accessible from your system's PATH.
2. The required Julia packages (SHA, JSON3) must be installed:
   In Julia REPL: `using Pkg; Pkg.add(["SHA", "JSON3"])`
3. The `juliacall` Python package must be installed:
   In terminal: `pip install juliacall`
"""

import os
from pathlib import Path
from typing import Union, List, Set, Dict, Tuple, Any

try:
    import juliacall
except ImportError:
    raise ImportError(
        "The 'juliacall' package is required. Please install it with 'pip install juliacall'."
    )

# --- Julia Module Loading ---

def _load_julia_module():
    """
    Loads the HllSets.jl module from the same directory as this Python file.
    This function is called once when the module is imported.
    """
    try:
        # Get the directory of this Python file
        current_dir = Path(__file__).parent
        julia_file_path = current_dir / "HllSets.jl"

        if not julia_file_path.exists():
            raise FileNotFoundError(
                f"Could not find HllSets.jl at {julia_file_path}. "
                "Ensure it is in the same directory as HLLSets.py"
            )

        # Include the Julia module. This makes it available as juliacall.Main.HllSets
        juliacall.Main.include(str(julia_file_path))
        
        # Verify that Julia dependencies are installed
        HllSets = juliacall.Main.HllSets
        juliacall.Main.eval("using HllSets")
        
        return HllSets

    except Exception as e:
        raise RuntimeError(f"Failed to load the HllSets.jl module: {e}")

# Load the Julia module when this script is first imported
try:
    jl_HllSets = _load_julia_module()
except Exception as e:
    # If loading fails, we create a dummy module to raise errors on use
    print(f"ERROR: Could not initialize HLLSets wrapper. {e}")
    jl_HllSets = None


# --- Python Wrapper Class ---

class HllSet:
    """
    A Python wrapper for the HllSets.HllSet Julia object.

    This class provides a Pythonic API for interacting with HyperLogLog sets,
    including adding elements, estimating cardinality, and performing set operations
    like union, intersection, and difference.

    Example:
        >>> hll1 = HllSet(p=10)
        >>> hll1.add("apple")
        >>> hll1.add("banana")
        >>> print(hll1.count)
        2
        >>> hll2 = HllSet(p=10)
        >>> hll2.add("banana")
        >>> hll2.add("cherry")
        >>> union_hll = hll1.union(hll2)
        >>> print(union_hll.count)
        3
    """

    def __init__(self, p: int = 10):
        """
        Initializes a new HllSet.

        Args:
            p (int): The precision of the HLL. Determines the number of registers
                     (2^p). Must be between 4 and 18. Default is 10.
        """
        if jl_HllSets is None:
            raise RuntimeError("HllSets.jl module could not be loaded.")
        self._jl_hll = jl_HllSets.HllSet(p)
        self.p = p

    @classmethod
    def _from_julia(cls, jl_hll_instance: Any):
        """
        Internal helper to create a Python HllSet from a Julia HllSet instance.
        This avoids re-calling the Julia constructor.
        """
        instance = cls.__new__(cls)  # Create instance without calling __init__
        instance._jl_hll = jl_hll_instance
        instance.p = jl_hll_instance.P # Access the type parameter P from Julia
        return instance

    # --- Core Operations ---

    def add(self, element: Any, seed: int = 0):
        """
        Adds an element to the HLL set.

        Args:
            element (Any): The element to add. Can be any hashable type.
            seed (int): An optional seed for the hash function.
        """
        self._jl_hll.add!(element, seed=seed)

    def add_all(self, elements: Union[List[Any], Set[Any]], seed: int = 0):
        """
        Adds a collection of elements to the HLL set.

        Args:
            elements (list or set): A collection of elements to add.
            seed (int): An optional seed for the hash function.
        """
        self._jl_hll.add!(elements, seed=seed)

    @property
    def count(self) -> int:
        """Estimates the cardinality (number of unique elements) in the set."""
        return int(self._jl_hll.count())

    @property
    def is_empty(self) -> bool:
        """Checks if the HLL set is empty."""
        return bool(self._jl_hll.isempty())

    @property
    def get_id(self) -> str:
        """Computes a stable SHA1 hash of the HLL set's contents."""
        return str(self._jl_hll.id())

    # --- Set Operations ---

    def union(self, other: 'HllSet') -> 'HllSet':
        """
        Computes the union of this HLL set with another.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            HllSet: A new HLL set representing the union.
        """
        new_jl_hll = self._jl_hll.union(other._jl_hll)
        return HllSet._from_julia(new_jl_hll)

    def union_inplace(self, other: 'HllSet') -> 'HllSet':
        """
        Modifies this HLL set in-place to be the union with another.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            HllSet: Returns self to allow for method chaining.
        """
        self._jl_hll.union!(other._jl_hll)
        
        return self

    def intersect(self, other: 'HllSet') -> 'HllSet':
        """
        Computes the intersection of this HLL set with another.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            HllSet: A new HLL set representing the intersection.
        """
        new_jl_hll = self._jl_hll.intersect(other._jl_hll)
        return HllSet._from_julia(new_jl_hll)

    def diff(self, other: 'HllSet') -> Dict[str, 'HllSet']:
        """
        Computes the set difference between this HLL set and another.

        Returns a dictionary containing three HllSets:
        - 'left_exclusive': Elements in self but not in other.
        - 'intersection': Elements in both self and other.
        - 'right_exclusive': Elements in other but not in self.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            dict: A dictionary of the three resulting HllSets.
        """
        # The Julia `diff` returns a named tuple, which juliacall converts to a Python object
        jl_result = self._jl_hll.diff(other._jl_hll)
        return {
            "left_exclusive": HllSet._from_julia(jl_result.DEL),
            "intersection": HllSet._from_julia(jl_result.RET),
            "right_exclusive": HllSet._from_julia(jl_result.NEW),
        }

    def symmetric_difference(self, other: 'HllSet') -> 'HllSet':
        """
        Computes the symmetric difference (XOR) of this HLL set with another.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            HllSet: A new HLL set representing the symmetric difference.
        """
        new_jl_hll = self._jl_hll.set_xor(other._jl_hll)
        return HllSet._from_julia(new_jl_hll)

    # --- Similarity Measures ---

    def jaccard_similarity(self, other: 'HllSet') -> float:
        """
        Computes the Jaccard similarity (as a percentage) with another HLL set.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            float: Jaccard similarity score between 0.0 and 100.0.
        """
        return float(self._jl_hll.match(other._jl_hll))

    def cosine_similarity(self, other: 'HllSet') -> float:
        """
        Computes the cosine similarity with another HLL set.

        Args:
            other (HllSet): The other HLL set.

        Returns:
            float: Cosine similarity score between 0.0 and 1.0.
        """
        return float(self._jl_hll.cosine(other._jl_hll))

    # --- Serialization ---

    def to_vector(self) -> List[int]:
        """
        Serializes the HLL set's internal registers to a list of integers.
        This is equivalent to the deprecated `dump` function in Julia.
        """
        return list(self._jl_hll.dump())

    @classmethod
    def from_vector(cls, data: List[int], p: int) -> 'HllSet':
        """
        Restores an HLL set from a list of integers created by `to_vector`.

        Args:
            data (list[int]): The list of register counts.
            p (int): The precision of the HLL set to be restored.

        Returns:
            HllSet: The restored HLL set.
        """
        hll = cls(p)
        hll._jl_hll.restore!(data)  # Ensure no invalid characters or encoding issues
        
        return hll

    # --- Python Magic Methods ---

    def __len__(self) -> int:
        """Allows using len(hll_set) to get the estimated count."""
        return self.count

    def __eq__(self, other: object) -> bool:
        """Checks if two HLL sets are identical (have the same registers)."""
        if not isinstance(other, HllSet):
            return NotImplemented
        return bool(self._jl_hll.isequal(other._jl_hll))

    def __repr__(self) -> str:
        """Provides a developer-friendly string representation of the HllSet."""
        return f"HllSet(p={self.p}, estimated_count={self.count})"


# --- Usage Example ---

if __name__ == '__main__':
    if jl_HllSets is None:
        print("Cannot run example: HllSets.jl module failed to load.")
    else:
        print("--- HLLSets Python Wrapper Example ---")

        # 1. Create two HLL sets
        hll_a = HllSet(p=12)
        hll_b = HllSet(p=12)

        # 2. Add elements to the sets
        data_a = ["apple", "banana", "cherry", "date", "elderberry"]
        data_b = ["banana", "cherry", "fig", "grape", "honeydew"]

        hll_a.add_all(data_a)
        hll_b.add_all(data_b)

        print(f"Set A created with {len(data_a)} elements. Estimated count: {hll_a.count}")
        print(f"Set B created with {len(data_b)} elements. Estimated count: {hll_b.count}")
        print("-" * 20)

        # 3. Perform set operations
        union_ab = hll_a.union(hll_b)
        print(f"Union of A and B. Estimated count: {union_ab.count}")

        intersection_ab = hll_a.intersect(hll_b)
        print(f"Intersection of A and B. Estimated count: {intersection_ab.count}")

        diff_result = hll_a.diff(hll_b)
        print(f"Items in A not in B: {diff_result['left_exclusive'].count}")
        print(f"Items in B not in A: {diff_result['right_exclusive'].count}")
        print("-" * 20)

        # 4. Calculate similarity
        jaccard = hll_a.jaccard_similarity(hll_b)
        print(f"Jaccard Similarity between A and B: {jaccard:.2f}%")
        
        cosine = hll_a.cosine_similarity(hll_b)
        print(f"Cosine Similarity between A and B: {cosine:.4f}")
        print("-" * 20)

        # 5. Serialization and Deserialization
        print(f"ID of Set A: {hll_a.get_id[:16]}...")
        
        serialized_data = hll_a.to_vector()
        print(f"Serialized Set A to a vector of length {len(serialized_data)}")
        
        restored_hll_a = HllSet.from_vector(serialized_data, p=12)
        print(f"Restored Set A is equal to original: {restored_hll_a == hll_a}")
        print(f"Restored Set A estimated count: {restored_hll_a.count}")
        print("-" * 20)

        # 6. Demonstrate __len__ and __repr__
        print(f"len(hll_a) returns: {len(hll_a)}")
        print(f"repr(hll_a) returns: {repr(hll_a)}")

File: deepseek_ocr/hllset/HllSets.jl
"""
MIT License

Copyright (c) 2023: Jakob Nybo Nissen.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

https://github.com/jakobnissen/Probably.jl/blob/master/src/hyperloglog/hyperloglog.jl

"""

"""
HyperLogLog Set Implementation with Extended Set Operations

This module provides a HyperLogLog (HLL) implementation with additional set operations
including union, intersection, difference, and complement operations.

Key Features:
- Efficient cardinality estimation using HLL algorithm
- Set operations between HLL sets
- Serialization/deserialization support
- Similarity measures (Jaccard, Cosine)
"""

module HllSets

    # NOTE: The user must have a `constants.jl` file with RAW_ARRAYS and BIAS_ARRAYS
    # and the JSON3.jl and SHA.jl packages installed.
    include("constants.jl")
    using SHA
    using JSON3

    export HllSet, add!, count, union, intersect, diff, isequal, isempty, id, delta, getbin, getzeros, maxidx, match, cosine, dump, restore, to_binary_tensor, flatten_tensor, tensor_to_string, string_to_tensor, binary_tensor_to_hllset, set_xor, set_comp, set_added, set_deleted

    struct HllSet{P}
        counts::Vector{UInt32}

        function HllSet{P}() where {P}
            isa(P, Integer) || throw(ArgumentError("P must be integer"))
            (P < 4 || P > 18) && throw(ArgumentError("P must be between 4 and 18"))
            new(fill(UInt32(0), 2^P))
        end
    end

    function HllSet(p::Int=10)
        return HllSet{p}()
    end

    # Core HLL Operations --------------------------------------------------------

    """
        add!(hll::HllSet{P}, x::Any; seed::UInt64=0)

    Add an element to the HLL set.
    """
    function add!(hll::HllSet{P}, x::Any; seed::UInt64 = 0x0) where {P}
        h = u_hash(x; seed=seed)
        bin = getbin(h; P=P)
        # The number of leading zeros + 1 is the rank to be stored
        idx = getzeros(h; P=P)
        if idx <= 32
            hll.counts[bin] |= (UInt32(1) << (idx - 1))
        end
    end

    function add!(hll::HllSet{P}, values::Union{Set, Vector}; seed::UInt64 = 0x0) where {P}
        for value in values
            add!(hll, value, seed=seed)
        end
    end    

    # Helper Functions ----------------------------------------------------------

    function _validate_compatible(x::HllSet{P}, y::HllSet{P}) where {P}
        length(x.counts) == length(y.counts) || 
            throw(ArgumentError("HLL sets must have same precision P"))
    end
    
    ### IMPROVED: Added a clear helper function
    """
        num_registers(hll::HllSet)

    Returns the number of registers (bins) in the HllSet, which is 2^P.
    """
    num_registers(::HllSet{P}) where {P} = 1 << P
    num_registers(::Type{HllSet{P}}) where {P} = 1 << P

    # Set Operations ------------------------------------------------------------

    """
        union(x::HllSet{P}, y::HllSet{P}) where {P}

    Compute union of two HLL sets.
    """
    function Base.union!(dest::HllSet{P}, src::HllSet{P}) where {P}
        _validate_compatible(dest, src)
        @inbounds for i in eachindex(dest.counts)
            ### IMPROVED: Removed redundant broadcasting dot
            dest.counts[i] |= src.counts[i]
        end
        return dest
    end

    function Base.union(x::HllSet{P}, y::HllSet{P}) where {P} 
        _validate_compatible(x, y)
        z = HllSet{P}()
        @inbounds for i in eachindex(x.counts)
            ### IMPROVED: Removed redundant broadcasting dot
            z.counts[i] = x.counts[i] | y.counts[i]
        end
        return z
    end

    """
        intersect(x::HllSet{P}, y::HllSet{P}) where {P}

    Compute intersection of two HLL sets.
    """
    function Base.intersect(x::HllSet{P}, y::HllSet{P}) where {P} 
        _validate_compatible(x, y)
        z = HllSet{P}()
        @inbounds for i in eachindex(x.counts)
            z.counts[i] = x.counts[i] & y.counts[i]
        end
        return z
    end

    """
        diff(hll_1::HllSet{P}, hll_2::HllSet{P}) where {P}

    Compute the set difference between two HLL sets.
    Returns a named tuple with:
    - `left_exclusive`: Elements in hll_1 but not in hll_2.
    - `intersection`: Elements in both hll_1 and hll_2.
    - `right_exclusive`: Elements in hll_2 but not in hll_1.
    """
    function Base.diff(hll_1::HllSet{P}, hll_2::HllSet{P}) where {P}
        _validate_compatible(hll_1, hll_2)
        
        # Elements in hll_1 but not in hll_2
        left_exclusive = set_comp(hll_1, hll_2)
        # Elements in hll_2 but not in hll_1
        right_exclusive = set_comp(hll_2, hll_1)
        # Elements in both
        intersection = intersect(hll_1, hll_2)

        return (left_exclusive = left_exclusive, intersection = intersection, right_exclusive = right_exclusive)
    end

    """
        set_comp(x::HllSet{P}, y::HllSet{P}) where {P}

    Compute the elements of `x` that are not in `y`. (i.e., `x \\ y`).
    """
    function set_comp(x::HllSet{P}, y::HllSet{P}) where {P} 
        _validate_compatible(x, y)
        z = HllSet{P}()
        @inbounds for i in eachindex(x.counts)
            ### IMPROVED: Simplified bitwise logic
            z.counts[i] = x.counts[i] & ~y.counts[i]
        end
        return z
    end

    """
        set_xor(x::HllSet{P}, y::HllSet{P}) where {P}

    Compute the symmetric difference between two HLL sets.
    """
    function set_xor(x::HllSet{P}, y::HllSet{P}) where {P} 
        _validate_compatible(x, y)
        z = HllSet{P}()
        @inbounds for i in eachindex(x.counts)
            ### IMPROVED: Removed redundant broadcasting dot
            z.counts[i] = xor(x.counts[i], y.counts[i])
        end
        return z
    end

    ### IMPROVED: Replaced copy!(src) with standard copy(src)
    """
        copy(src::HllSet{P})

    Create a new, independent copy of an HllSet.
    """
    function Base.copy(src::HllSet{P}) where {P}
        dest = HllSet{P}()
        copyto!(dest.counts, src.counts)
        return dest
    end

    """
        copy!(dest::HllSet{P}, src::HllSet{P})

    Copy the contents of `src` into `dest`. `dest` must be a pre-existing HllSet.
    """
    function Base.copy!(dest::HllSet{P}, src::HllSet{P}) where {P}
        _validate_compatible(dest, src)
        copyto!(dest.counts, src.counts)
        return dest
    end

    """
        isequal(x::HllSet{P}, y::HllSet{P}) where {P}

    Check if two HLL sets are equal (i.e., have identical registers).
    """    
    function Base.isequal(x::HllSet{P}, y::HllSet{P}) where {P} 
        _validate_compatible(x, y)
        return x.counts == y.counts
    end    

    Base.isempty(x::HllSet{P}) where {P} = all(iszero, x.counts)   

    """
        count(x::HllSet{P}) where {P}

    Estimate the cardinality of the HLL set.
    """
    ### IMPROVED: Fixed critical bug in cardinality calculation
    function Base.count(x::HllSet{P}) where {P}
        m = num_registers(x) # Correctly get number of registers
        harmonic_sum = sum(inv(1 << maxidx(i)) for i in x.counts)
        harmonic_mean = m / harmonic_sum
        
        biased_estimate = α(x) * m * harmonic_mean
        return round(Int, biased_estimate - bias(x, biased_estimate))
    end

    """
        Set of helper functions for cardinality estimation.
    """
    ### IMPROVED: Now uses UInt64 and is much faster
    """
        getbin(x::UInt64; P::Int)

    Extracts the top `P` bits from a 64-bit hash `x` to determine the register index.
    """
    function getbin(x::UInt64; P::Int=10)
        return Int((x >>> (64 - P)) + 1) # +1 for 1-based indexing
    end

    ### IMPROVED: Now uses UInt64
    """
        getzeros(x::UInt64; P::Int)

    Counts the number of leading zeros in the lower `64-P` bits of the hash `x`.
    """
    function getzeros(x::UInt64; P::Int=10)
        # Mask to get the lower bits used for zero-counting
        lower_bits_mask = (UInt64(1) << (64 - P)) - 1
        # Count trailing zeros of the lower bits. If all are zero, count is 64-P.
        # We add 1 because HLL rank is 1-based.
        return trailing_zeros(x & lower_bits_mask) + 1
    end

    ### IMPROVED: Simplified logic
    α(x::HllSet{P}) where {P} = if P == 4; 0.673
        elseif P == 5; 0.697
        elseif P == 6; 0.709
        else; 0.7213 / (1 + 1.079 / num_registers(x))
        end 
    
    function bias(::HllSet{P}, biased_estimate) where {P}
        if P < 4 || P > 18
            error("We only have bias estimates for P ∈ 4:18")
        end
        rawarray = @inbounds RAW_ARRAYS[P - 3]
        biasarray = @inbounds BIAS_ARRAYS[P - 3]
        firstindex = searchsortedfirst(rawarray, biased_estimate)
        if firstindex == length(rawarray) + 1
            return 0.0
        elseif firstindex == 1
            return @inbounds biasarray[1]
        else
            x1, x2 = @inbounds rawarray[firstindex - 1], @inbounds rawarray[firstindex]
            y1, y2 = @inbounds biasarray[firstindex - 1], @inbounds biasarray[firstindex]
            delta = @fastmath (biased_estimate - x1) / (x2 - x1)
            return y1 + delta * (y2 - y1)
        end
    end

    ### IMPROVED: Simplified implementation
    """
        maxidx(x::UInt32)

    Returns the position of the most significant bit (MSB) in a 32-bit register.
    This corresponds to the maximum rank stored in that register.
    """
    function maxidx(x::UInt32)        
        # If x is 0, there are no bits set, so the rank is 0.
        # Otherwise, find the position of the MSB.
        return x == 0 ? 0 : (sizeof(UInt32) * 8 - leading_zeros(x))
    end

    # Match Operations ------------------------------------------------------------

    """
        match(x::HllSet{P}, y::HllSet{P}) where {P}
    Compute the Jaccard similarity between two HLL sets as a percentage (0-100).
    """
    function Base.match(x::HllSet{P}, y::HllSet{P}) where {P}
        _validate_compatible(x, y)
        
        count_u = count(union(x, y))
        # Avoid division by zero if both sets are empty
        count_u == 0 && return 100

        count_i = count(intersect(x, y))
        return round(Int64, (count_i / count_u) * 100)
    end

    """
        cosine(hll_1::HllSet{P}, hll_2::HllSet{P}) where {P}
    Compute the cosine similarity between two HLL sets.
    """
    function cosine(hll_1::HllSet{P}, hll_2::HllSet{P}) where {P}
        _validate_compatible(hll_1, hll_2)

        v1 = hll_1.counts
        v2 = hll_2.counts
        norm_v1 = norm(v1)
        norm_v2 = norm(v2)
        
        if norm_v1 == 0 || norm_v2 == 0
            return 0.0
        end
        return dot(v1, v2) / (norm_v1 * norm_v2)
    end

    # Serialization and Deserialization ----------------------------------------

    """
        to_binary_tensor(hll::HllSet{P}) where {P}
    Convert the HLL set to a binary tensor (Matrix of Bools).
    """
    function to_binary_tensor(hll::HllSet{P}) where {P}
        m = num_registers(hll)
        tensor = falses(m, 32)
        for i in 1:m
            # reinterpret is a fast way to get bits
            bits = reinterpret(Bool, hll.counts[i])
            # Note: bitstring is little-endian, so we reverse to match intuition
            tensor[i, :] .= reverse(bits)
        end
        return tensor
    end

    """
        flatten_tensor(tensor::Array{Bool, 2})
    Flatten the binary tensor to a 1D array.
    """
    flatten_tensor(tensor::Array{Bool, 2}) = vec(tensor)

    """
        tensor_to_string(flattened_tensor::Vector{Bool})
    Convert a flattened binary tensor to a string representation.
    """
    tensor_to_string(flattened_tensor::Vector{Bool}) = join([b ? "1" : "0" for b in flattened_tensor], "")

    """
        string_to_tensor(str::String, P::Int=10)
    Convert a string representation of a binary tensor to a 2D array.
    """
    function string_to_tensor(str::String, P::Int=10)
        m = 2^P
        expected_len = m * 32
        if length(str) != expected_len
            str = rpad(str, expected_len, "0")
        end
        return reshape([c == '1' for c in str], (m, 32))
    end

    """
        binary_tensor_to_hllset(tensor::Array{Bool, 2}, P::Int=10)
    Convert a binary tensor to an HLL set.
    """
    function binary_tensor_to_hllset(tensor::Array{Bool, 2}, P::Int=10)
        hll = HllSet{P}()
        for i in 1:num_registers(hll)
            # Reverse bits back to little-endian for reinterpret
            bits_str = join(reverse(tensor[i, :]))
            hll.counts[i] = parse(UInt32, bits_str, base=2)
        end
        return hll
    end   
   
    # HllSet ID ---------------------------------------------------
    
    """
        id(x::HllSet{P}) where {P}
    Compute a stable SHA1 hash of the HllSet's contents, useful for identification.
    """
    function id(x::HllSet{P}) where {P}
        isnothing(x) && return nothing
        bytearray = reinterpret(UInt8, x.counts)
        hash_value = SHA.sha1(bytearray)
        return SHA.bytes2hex(hash_value)
    end

    ### IMPROVED: Now a safe, efficient alias for id()
    function sha1(x::HllSet{P}) where {P}
        @warn "sha1(hll) is deprecated in favor of id(hll) for clarity." maxlog=1
        return id(x)
    end

    ### IMPROVED: Now uses UInt64 and warns about stability
    """
        u_hash(x; seed::UInt64 = 0x0)

    Computes a 64-bit unsigned hash for an element `x`.
    **WARNING:** Uses Julia's built-in `hash`, which is not guaranteed to be stable
    across different Julia versions. For applications requiring serialization and
    deserialization, a stable hash like SHA-256 should be used instead.
    """
    function u_hash(x; seed::UInt64 = 0x0) 
        h = reinterpret(UInt64, hash(x)) ⊻ seed
        return h
    end

    # Overload the show function to print the HllSet --------------------------------------------------
    Base.show(io::IO, x::HllSet{P}) where {P} = print(io, "HllSet{$(P)}($(count(x)) estimated elements)")

    ### IMPROVED: Removed confusing sizeof overloads
    # Base.sizeof(::Type{HllSet{P}}) where {P} = 1 << P
    # Base.sizeof(x::HllSet{P}) where {P} = sizeof(typeof(x))

    # Depricated Functions --------------------------------------------------
    
    function Base.dump(x::HllSet{P}) where {P}
        @warn "dump(hll) is deprecated, access hll.counts directly." maxlog=1
        return x.counts
    end

    ### IMPROVED: Fixed bugs and used fast copyto!
    function restore!(z::HllSet{P}, x::Vector{UInt32}) where {P} 
        if length(x) != num_registers(z)
            error("The length of the vector must be equal to the number of HllSet registers.")
        end        
        copyto!(z.counts, x)
        return z
    end

    function restore!(z::HllSet{P}, x::String) where {P}
        dataset = JSON3.read(x, Vector{UInt32})
        if length(dataset) != num_registers(z)
            error("JSON data length does not match HllSet size.")
        end
        copyto!(z.counts, dataset)
        return z
    end 

end